src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -491,6 +492,46 @@ def _compile_hermetic(self, jvm_options, ctx, classes_dir, jar_file,
     # TODO: This should probably return a ClasspathEntry rather than a Digest
     return res.output_directory_digest
 
+  def _compute_scratch_inputs(self, classes_dir, relpath_to_analysis, jar_file):
+    """
+    Compute for the scratch inputs for ExecuteProcessRequest.
+
+    :param classes_dir: relative path to classes dir from buildroot
+    :param relpath_to_analysis: relative path to zinc analysis file from buildroot
+    :param jar_file: relative path to z.jar from buildroot
+    :return: digest of merged analysis file and loose class files.
+    """
+    if not os.path.exists(relpath_to_analysis):
+      return EMPTY_DIRECTORY_DIGEST
+
+    def _get_analysis_snaoshot():
+      _analysis_snapshot, = self.context._scheduler.capture_snapshots([
+        PathGlobsAndRoot(
+          PathGlobs([relpath_to_analysis]),
+          get_buildroot(),
+        ),
+      ])
+      return _analysis_snapshot
+
+    def _get_classes_dir_snapshot():
+      if self.get_options().use_classpath_jars:
+        with zipfile.ZipFile(jar_file, 'r') as zip_ref:
```

Comment:
`jar_file` ideally _shouldn't_ exist on disk (and we have a ticket to make that the case that we'll get to at some point); if this is necessary, we should probably materialize it from Digest, or introduce a new "unzip_from_digest" intrinsic or something... But see also my questions about whether this is necessary

Reply:
It is not clear to me how to get current target's `z.jar` from digest.

I can add a check on `z.jar` first before unzipping to make it safer if needed.

Reply:
Yeah I think it will be cleaner if we do so in the zinc wrapper, but we can follow up.

Reply:
https://github.com/pantsbuild/pants/issues/8331